pandas/io/common.py

```diff
@@ -792,6 +792,10 @@ def get_handle(
             isinstance(ioargs.filepath_or_buffer, str) or ioargs.should_close
         )
 
+        if memory_map and encoding == "utf-8":
+            # c-engine can deal with str&bytes: no need to decode twice
+            handle.read = handles[-2].read
```

Comment:
Without this ugly line (skip TextIOWrapper.read and directly use mmap.read), this PR would effectively revert #43787.

Reply:
It might be nicer to inherit from `TextIOWrapper` and overwrite `read`?